src/server/acl/acl_family.cc

```diff
@@ -188,12 +188,44 @@ void AclFamily::SetUser(CmdArgList args, ConnectionContext* cntx) {
     ServerState::tlocal()->user_registry->MaybeAddAndUpdate(username, std::move(req));
     auto cred = ServerState::tlocal()->user_registry->GetCredentials(username);
     StreamUpdatesToAllProactorConnections(username, cred.acl_categories);
-    (*cntx)->SendOk();
+    cntx->SendOk();
   };
 
   std::visit(Overloaded{error_case, update_case}, std::move(req));
 }
 
+void AclFamily::EvictOpenConnectionsOnAllProactors(std::string_view user) {
+  auto close_cb = [user]([[maybe_unused]] size_t id, util::Connection* conn) {
+    DCHECK(conn);
+    auto connection = static_cast<facade::Connection*>(conn);
+    auto ctx = static_cast<ConnectionContext*>(connection->cntx());
+    if (ctx && ctx->authed_username == user) {
+      // ctx->CancelBlocking();
```

Comment:
How do we abort already existing transactions? There seems to be done in two ways: 

1. For blocking transactions via `CancelBlocking`
2. From `Transaction::ScheduleInternal::632` which I am not 100% sure how it works

Is there a generic way to do this? If not I will need to investigate and figure the details.

Reply:
> We can keep it as a to-do for the near future. We might need to add an abort flag to the transaction and then check it in the right places & remove transaction from scheduled transactions queue

I can take care of that.

> my main concern for now is that if we release the `connection_context` (since we are shutting down the connection) is that we will end up accessing a `released connection context *` which means we flew to the realm of UB :D

Reply:
honestly perfect opportunity for me to dive into our transactional framework (was eagerly waiting for that moment :eyes: )

Reply:
Yes you are right because out transactional framework has no notion of a `delta`. We lock the keys and we start executing. 

For commands that operate on a single key, that's easy, since it's a single delta that is VERY easy to handle. For commands with big changes such as lua scripts, this is an issue and probably a performance hit. 

I don't think there is a good solution for this but I will need to think this, maybe for now it suffices that we don't immediately evict connections but rather give them time to wrap up their transactions and then exit. That at least will guarantee proper lifetime semantics of the underline contexts as well as stop the users for issuing more commands after they have been kicked out :stuck_out_tongue:

Reply:
At this point in the discussion, I think it's worth it to just swallow the 'soft delete' pill and leak the user objects indefinitely.

Reply:
+1, it's not something we should do now. What is important now though is to assert that if we kill a connection the underline context won't hang or be accessed by any transaction. We don't want to be accessing nullptr's (since the connection itself holds a pointer to the underline context)

Reply:
Actually as of a second thought I don't know why you complicate things so much.

> If we add a mechanism for aborting commands we'd also need a rollback mechanism, no?

No. Its nearly impossible. We should see what Redis does and choose the simplest approach.

ACL DELUSER is most likely handled just as a regular command, so because redis is single threaded, it executes after all preceding commands of the same user (same linear order). How do we achieve this 'cut' on our own? Lets make ACL DELUSER a global transaction - it guarantees us no commands are executing (We _could_ guarantee no commands of the _same_ user are running, which is more efficient, but the global transaction is surely enough and doesn't require additional mechanisms)

Reply:
@dranikpg I discussed this with Roy. There is a better solution. We `shutdown` the context, this won't cause any issue, it just closes the socket. If the transaction is executing that's fine, we let it finish. If it's in the middle of `multi/exec` we handle this by polling the is_shutting down flag. Otherwise, if it's a script we let it finish, it started before we kicked the user out of the system so the behaviour is still correct.

Reply:
> Otherwise, if it's a script we let it finish, it started before we kicked the user out of the system so the behaviour is still correct.

But multi/exec also started before....

Reply:
hmm actually that's right!